// @zerodevapp/sdk/dist/src/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getERC1155Contract = exports.getERC20Contract = exports.getERC721Contract = exports.fixSignedData = exports.hexifyUserOp = exports.getRpcUrl = void 0;
const ethers_1 = require("ethers");
const utils_1 = require("ethers/lib/utils");
const constants = __importStar(require("./constants"));
const getRpcUrl = (chainId) => {
    return `https://${constants.CHAIN_ID_TO_INFURA_NAME[chainId]}.infura.io/v3/${constants.INFURA_API_KEY}`;
};
exports.getRpcUrl = getRpcUrl;
const hexifyUserOp = (resolvedUserOp) => {
    return Object.keys(resolvedUserOp)
        .map((key) => {
        let val = resolvedUserOp[key];
        if (typeof val !== 'string' || !val.startsWith('0x')) {
            val = (0, utils_1.hexValue)(val);
        }
        return [key, val];
    })
        .reduce((set, [k, v]) => (Object.assign(Object.assign({}, set), { [k]: v })), {});
};
exports.hexifyUserOp = hexifyUserOp;
// Some signers do not return signed data with 0x prefix, which makes it 
// an invalid hex byte string.  So we first check if it's a hex string,
// and if it's not, we prepend 0x and check if it's a valid hex string.
// If it's still not, we throw an error.
const fixSignedData = (sig) => {
    if (ethers_1.ethers.utils.isHexString(sig)) {
        return sig;
    }
    const fixedSig = `0x${sig}`;
    if (ethers_1.ethers.utils.isHexString(fixedSig)) {
        return fixedSig;
    }
    throw new Error('Invalid signed data ' + sig);
};
exports.fixSignedData = fixSignedData;
const getERC721Contract = (provider, address) => {
    return new ethers_1.Contract(address, constants.ERC721_ABI, provider);
};
exports.getERC721Contract = getERC721Contract;
const getERC20Contract = (provider, address) => {
    return new ethers_1.Contract(address, constants.ERC20_ABI, provider);
};
exports.getERC20Contract = getERC20Contract;
const getERC1155Contract = (provider, address) => {
    return new ethers_1.Contract(address, constants.ERC1155_ABI, provider);
};
exports.getERC1155Contract = getERC1155Contract;
